File: 0x02/stat.py
#!/bin/env python3
# NEED TO CONSIDER SPECIAL CHARS + ELLIPSIS!!!
import sys

# OPEN FILE
try:
    filename = sys.argv[1]
    f = open(filename)
except:
    print('ERROR: please give me a file')
    sys.exit(1)

# GEN DATA
d_words = {}
with f:
    for line in f:
       for word in map(str.lower, line.split()):
            d_words[word] = 1 if word not in d_words else d_words[word]+1
d_lengths = {}
for word, count in d_words.items():
    d_lengths[len(word)] = count if len(word) not in d_lengths else d_lengths[len(word)] + count

# GEN STATS
tot = sum(d_lengths.values())
if tot == 0:
    print('ERROR: file is empty')
    sys.exit(1)
avg = sum(k*v for k,v in d_lengths.items()) // sum(d_words.values())
stat = sorted(d_lengths.items())
popular = list(sorted(d_words.items(), key=lambda t:t[1], reverse=True))[:10]

# OUTPUT
print(f'Total Words: {tot}\n')
print(f'Average Length: {avg}\n')
print('Top Lengths:')
for l,n in stat:
    print(f'{l:3} chars => {n:5}')
print('\nTop Words:')
for i,(w,n) in enumerate(popular, start=1):
    print(f'{i:2}. {w:10} => {n:5}')

File: 0x02/stat.hs
import Data.Char 
import qualified Data.Map.Lazy as Map
import Data.List
--import Control.Exception
import Text.Printf
import System.Environment
import Control.Monad
import System.Exit

main :: IO ()
main = do
    -- OPEN FILE
    argv <- getArgs
    when (null argv) exitFailure
    file <- readFile (head argv)


    -- GENEREATE STRUCTURES
    let d_words  = Map.fromListWith (+) [(word,1) | word<-words (map toLower file)]
    let d_lengths = Map.fromListWith (+) [(length word, num) | (word, num) <- Map.assocs d_words]

    -- GENEREATE STATISTICS
    let tot = Map.foldl (+) 0 d_lengths
    let avg =  (Map.foldlWithKey (\acc key val -> acc+val*key) 0 d_lengths) `quot` tot
    let stat = Map.assocs d_lengths
    let _popular = sortBy (\a b-> (flip compare) (snd a) (snd b)) (Map.assocs d_words)
    let popular = take 10 _popular

    -- OUTPUT
    putStrLn $ printf "Total Words: %d\n" tot
    putStrLn $ printf "Average Length: %d\n" avg
    putStrLn $ "Top Lengths:\n" ++ unlines (map (\(a,b) -> printf "%3d chars => %5d" a b) stat)
    putStrLn $ "Top Lengths:\n" ++ unlines (map (\(i,(a,b)) -> printf "%2d %10s => %5d" i a b) (zip ([1..]::[Int]) popular))

